src/history/hs_rec.c

```diff
@@ -878,15 +702,19 @@ __hs_fixup_out_of_order_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
      * to keep doing "next" until we've got a key greater than the one we attempted to position
      * ourselves with.
      */
-    for (; ret == 0; ret = __wt_hs_cursor_next(session, hs_cursor)) {
+    for (; ret == 0; ret = hs_cursor->next(hs_cursor)) {
         /*
          * Prior to getting here, we've done a "search near" on our key for the timestamp we're
          * inserting and then a "next". In the regular case, our cursor will be positioned on the
          * next key and we'll break out of the first iteration in one of the conditions below.
          */
         WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, &hs_key, &hs_ts, &hs_counter));
-        WT_ERR(__wt_compare(session, NULL, &hs_cursor->key, srch_key, &cmp));
-        if (cmp > 0)
+        if (hs_btree_id != btree->id)
+            break;
+        if (hs_ts > ts)
+            break;
+        WT_ERR(__wt_compare(session, NULL, &hs_key, key, &cmp));
```

Comment:
Same for key. We can make this an assert instead.

Reply:
Fixed.